{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be96dc3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "#function to generate hashtable with kmer counts\n",
    "def count_kmers(sequence, hashtable, kmer_length, ignore_directionality, filter_kmer):\n",
    "    kmer = 6\n",
    "    for i in range(len(sequence)-kmer+1):\n",
    "        end = kmer + i\n",
    "        kmer_key = sequence[i:end]\n",
    "        hashtable[kmer_key] = hashtable.get(kmer_key, 0) + 1\n",
    "        #if ignore_directionality:\n",
    "            #kmer_key = ignore_directionality(sequence,kmer)\n",
    "        \n",
    "    #print(hashtable)\n",
    "\n",
    "# output file should be from previous commands in our program and contains calculated kmers and their frequencies\n",
    "# threshold parameter is an integer value given by the user. Any kmer frequency lower than the threshold will be dropped\n",
    "def filter_low_freq(outputfile, threshold):\n",
    "    kmer_counts = {}\n",
    "    \n",
    "    with open(outputfile) as outputfile:\n",
    "        outputlist = outputfile.read().splitlines()\n",
    "    for x in outputlist: \n",
    "        mylist = x.split('\\t',2)\n",
    "        \n",
    "        #print(mylist[0])\n",
    "        print(threshold)\n",
    "        print(mylist[1])\n",
    "        if(int(mylist[1]) < int(threshold)):\n",
    "            kmer_counts[mylist[0]] = int(mylist[1])\n",
    "            \n",
    "    with open(outputfile, \"w\") as out_file:\n",
    "        for kmer, count in kmer_counts.items():\n",
    "            outputfile.write(f\"{kmer}\\t{count}\\n\")\n",
    "    print(kmer_counts)\n",
    "    \n",
    "#function to read in input sequence from fasta file and output\n",
    "\n",
    "def kmer(input_fa, kmer_length, output, directionality, filter_kmer):\n",
    "\n",
    "    seq_file = input_fa\n",
    "\n",
    "    #implement hashtable using dictionary\n",
    "    hashtable = {}\n",
    "\n",
    "    for sequ in SeqIO.parse(open(seq_file),'fasta'):\n",
    "        sequence = str(sequ.seq)\n",
    "\n",
    "    #print(sequence)\n",
    "\n",
    "    count_kmers(sequence, hashtable, kmer_length, True, True)\n",
    "    \n",
    "    #output the kmer counts in a txt file\n",
    "    output_txt = output\n",
    "    with open(output_txt, \"w\") as out_file:\n",
    "        for kmer, count in hashtable.items():\n",
    "                out_file.write(f\"{kmer}\\t{count}\\n\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#function to generate reverse complement if directionality is ignored(still working on making this work)\n",
    "def ignore_directionality(sequence, kmer):\n",
    "    nucleotides = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "    #reverse_complement = ''.join(nucleotides[key] for key in reversed(sequence))\n",
    "    hashtable2 = {}\n",
    "    \n",
    "    for i in range(len(sequence)-kmer+1):\n",
    "        end = kmer + i\n",
    "        kmer_key = sequence[i:end]\n",
    "        kmer_complement = ''.join(nucleotides[key] for key in reversed(kmer_key))\n",
    "        print(kmer_complement)\n",
    "        hashtable2[kmer_key] = hashtable2.get(kmer_key, 0) + 1\n",
    "        hashtable2[kmer_complement] = hashtable2.get(kmer_complement, 0) + 1\n",
    "\n",
    "\n",
    "\n",
    "seq_file = \"example_genome.fa\"\n",
    "output_txt = \"output.txt\"\n",
    "kmer(seq_file, 6, output_txt, True, True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
